srt-protocol/src/connection.rs

```diff
@@ -49,3 +56,378 @@ pub struct ConnectionSettings {
+pub enum Action {
+    ReleaseData((Instant, Bytes)),
+    SendPacket((Packet, SocketAddr)),
+    WaitForData(Duration),
+    Close,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Input {
+    Data(Option<(Instant, Bytes)>),
+    Packet(Option<(Packet, SocketAddr)>),
+    DataReleased,
+    PacketSent,
+    Timer,
+}
+
+impl DuplexConnection {
+    pub fn new(connection: Connection) -> DuplexConnection {
+        let settings = connection.settings;
+        DuplexConnection {
+            status: ConnectionStatus::Open(settings.send_tsbpd_latency),
+            settings: settings.clone(),
+            receiver: Receiver::new(settings.clone(), connection.handshake.clone()),
+            sender: Sender::new(settings, connection.handshake),
+        }
+    }
+
+    pub fn handle_input(&mut self, now: Instant, input: Input) -> Action {
+        debug!(
+            "{:?}|{:?}|input - {:?}",
+            TimeSpan::from_interval(self.settings.socket_start_time, now),
+            self.settings.local_sockid,
+            input
+        );
+
+        match input {
+            Input::Data(Some(data)) => self.sender.handle_data(data, now),
+            Input::Data(None) => self.handle_data_stream_close(now),
+            Input::Packet(Some(packet)) => self.handle_packet(now, packet),
+            Input::Packet(None) => self.handle_socket_close(),
+            _ => {}
+        };
+
+        let action = if !self.ensure_open(now) {
+            Action::Close
+        } else if let Some(data) = self.next_data(now) {
+            Action::ReleaseData(data)
+        } else if let Some(packet) = self.next_packet() {
+            Action::SendPacket(packet)
+        } else {
+            Action::WaitForData(self.next_timer(now) - now)
+        };
+
+        debug!(
+            "{:?}|{:?}|action - {:?}",
+            TimeSpan::from_interval(self.settings.socket_start_time, now),
+            self.settings.local_sockid,
+            action
+        );
+
+        action
+    }
+
+    pub fn is_open(&self) -> bool {
+        !(self.status == ConnectionStatus::Closed
+            || (!self.sender.is_open() && self.receiver.is_flushed())
+            || (!self.receiver.is_open() && self.sender.is_flushed()))
+    }
+
+    pub fn next_packet(&mut self) -> Option<(Packet, SocketAddr)> {
+        self.sender
+            .next_packet()
+            .or_else(|| self.receiver.next_packet())
+    }
+
+    pub fn next_data(&mut self, now: Instant) -> Option<(Instant, Bytes)> {
+        self.receiver.next_data(now)
+    }
+
+    pub fn next_timer(&self, now: Instant) -> Instant {
+        min(self.sender.next_timer(now), self.receiver.next_timer(now))
+    }
+
+    pub fn ensure_open(&mut self, now: Instant) -> bool {
+        if !self.is_open() {
+            false
+        } else {
+            self.check_timers(now);
+            true
+        }
+    }
+
+    pub fn check_timers(&mut self, now: Instant) -> Instant {
```

Comment:
We should either have all the check_timers calls be `Instant -> Instant` or `Instant -> ()` instead of some of one some of the other. I think `Instant -> Instant` is the better way to go, as it ensures that you call check_timers and not just next_timer.